flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java

```diff
@@ -0,0 +1,235 @@
+
+    @Parameterized.Parameters
+    public static Object[][] data() {
+        return new Object[][] {
+            {new ResultSubpartitionRecoveredChannelStateHandlerTester()},
+            {new InputChannelRecoveredStateHandlerTester()},
+        };
+    }
+
+    public RecoveredChannelStateHandlerTest(
+            RecoveredChannelStateHandlerTester<Info, Context> stateHandlerTester) {
+        this.stateHandlerTester = stateHandlerTester;
+    }
+
+    @Test
+    public void testRecycleBufferBeforeRecoverWasCalled() throws Exception {
+        // given: Initialized state handler.
+        stateHandlerTester.init();
+
+        Info channelInfo = stateHandlerTester.getInfo();
+
+        // when: Request the buffer.
+        RecoveredChannelStateHandler.BufferWithContext<Context> buffer =
+                stateHandlerTester.getBuffer(channelInfo);
+
+        // and: Recycle buffer outside.
+        buffer.buffer.recycle();
+
+        // then: Buffer should be recycled the same times as it was retained.
+        stateHandlerTester.assertState();
+    }
+
+    @Test
+    public void testRecycleBufferAfterRecoverWasCalled() throws Exception {
+        // given: Initialized state handler.
+        stateHandlerTester.init();
+
+        Info channelInfo = stateHandlerTester.getInfo();
+
+        // when: Request the buffer.
+        RecoveredChannelStateHandler.BufferWithContext<Context> buffer =
+                stateHandlerTester.getBuffer(channelInfo);
+
+        // and: Pass the buffer to recovery.
+        stateHandlerTester.recover(channelInfo, 0, buffer.context);
+
+        // and: Recycle buffer outside.
+        buffer.buffer.recycle();
+
+        // then: Buffer should be recycled the same times as it was retained.
+        stateHandlerTester.assertState();
+    }
+
+    static class ResultSubpartitionRecoveredChannelStateHandlerTester
+            extends RecoveredChannelStateHandlerTester<ResultSubpartitionInfo, BufferBuilder> {
+        AtomicInteger recycledMemorySegments;
+        ResultPartition partition;
+
+        @Override
+        ResultSubpartitionInfo getInfo() {
+            return new ResultSubpartitionInfo(0, 0);
+        }
+
+        @Override
+        ResultSubpartitionRecoveredStateHandler createInstanceForTest() throws IOException {
+            // given: Result partition with recycling counter.
+            recycledMemorySegments = new AtomicInteger();
+            partition =
+                    new ResultPartitionBuilder()
+                            .setNetworkBufferPool(
+                                    new NetworkBufferPool(3, 32 * 1024) {
+                                        @Override
+                                        public void recycle(MemorySegment segment) {
+                                            recycledMemorySegments.incrementAndGet();
+                                            super.recycle(segment);
+                                        }
+                                    })
```

Comment:
Why not use plain simple `NetworkBufferPool` and just assert at the end the number of available segments instead of counting how many times `recycle` was called? Frankly, the test shouldn't care how many times `recycle` was called. Just that the buffers were returned to the pool.